### PR TITLE
[BRAIN-1527] feat: shadow layer

### PR DESCRIPTION
**Pull request by grieneis**

### CODE REVIEW

File: costmap_2d/include/costmap_2d/shadow_layer.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Daniel Grieneisen
+ *********************************************************************/
+#ifndef COSTMAP_2D_OBSTACLE_LAYER_H_
```
Comment (mrousseau65) [inline costmap_2d/include/costmap_2d/shadow_layer.h]: These need to be changed (or use #pragma once) to COSTMAP_2D_SHADOW_LAYER_

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/*********************************************************************
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2016, 6 River Systems
```
Comment (mrousseau65) [inline base_local_planner/src/shadow_speed_limiter.cpp]: Welcome to 2017. :)

File: costmap_2d/plugins/shadow_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+{
+  dsrv_ = new dynamic_reconfigure::Server<costmap_2d::ShadowPluginConfig>(nh);
+  dynamic_reconfigure::Server<costmap_2d::ShadowPluginConfig>::CallbackType cb = boost::bind(
+      &ShadowLayer::reconfigureCB, this, _1, _2);
+  dsrv_->setCallback(cb);
+}
+
+ShadowLayer::~ShadowLayer()
+{
+  if (dsrv_)
```
Comment (mrousseau65) [inline costmap_2d/plugins/shadow_layer.cpp]: the delete operator has a nullptr check internally, so this isn't technically needed.

File: costmap_2d/plugins/shadow_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+  double angle_start = -shadow_scan_half_angle_ + robot_yaw;
+  double angle_end = shadow_scan_half_angle_ + robot_yaw;
+  for (double angle = angle_start; 
+    angle <= angle_end; 
+    angle += shadow_scan_angular_resolution_)
+  {
+    double x = cos(angle) * shadow_scan_range_ + robot_x;
+    double y = sin(angle) * shadow_scan_range_ + robot_y;
+    observation_cloud.push_back(pcl::PointXYZ(x, y, 0.0)
+    );
```
Comment (mrousseau65) [inline costmap_2d/plugins/shadow_layer.cpp]: accidental carriage return?

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  double pose_yaw = tf::getYaw(pose.pose.orientation);
+  pose.pose.position.x += cos(pose_yaw) * params_.forward_offset_;
+  pose.pose.position.y += sin(pose_yaw) * params_.forward_offset_;
+
+  map_grid_.setUnadjustedGoal(*costmap_, pose);
+
+
+  // unsigned int px, py;
+  // if (!costmap_->worldToMap(current_pose_as_array_[0].pose.position.x, current_pose_as_array_[0].pose.position.y, px, py)) 
+  // {
```
Comment (dantwinkler) [inline base_local_planner/src/shadow_speed_limiter.cpp]: Remove.

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  //   ROS_WARN("Uhhhh...");
+  //   //we're off the map
+  // }
+  // ROS_DEBUG("Cost at goal: %f.  Goal %f, %f", map_grid_(px,py).target_dist, map_grid_.goal_x_,map_grid_.goal_y_);
+
+
+  // Find nearest object via brushfire distance
+
+  for (auto obj : (*objects_))
```
Comment (dantwinkler) [inline base_local_planner/src/shadow_speed_limiter.cpp]: Should be const &

File: costmap_2d/cfg/ShadowPlugin.cfg [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+gen.add("enabled", bool_t, 0, "Whether to apply this plugin or not", True)
+gen.add("publish_shadow_objects", bool_t, 0, "Whether to publish for visualization", True)
+
+gen.add("shadow_scan_range", double_t, 0, "Maximum range to check for scan shadows", 3.0, 0, 20)
+gen.add("shadow_scan_angular_resolution", double_t, 0, "Angular resolution of rays to check for shadows", 0.005, 0, 1.0)
+gen.add("shadow_scan_half_angle", double_t, 0, "Half width of scan to check for shadows", 1.57, 0, 3.14)
+gen.add("min_shadow_size", double_t, 0, "Radius of object that can fit in shadow to declare it shadowed", 0.15, 0, 2.0) 
+gen.add("sensor_x_offset", double_t, 0, "The offset in the x axis of the sensors", 0.5, -1.5, 1.5) 
+
+#gen.add("max_obstacle_range", double_t, 0, "The default maximum distance from the robot at which an obstacle will be inserted into the cost map in meters.", 2.5, 0, 50)
```
Comment (dantwinkler) [inline costmap_2d/cfg/ShadowPlugin.cfg]: Remove?

File: costmap_2d/include/costmap_2d/shadow_layer.h [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+
+  class ShadowMarker
+  {
+  public:
+    ShadowMarker(unsigned char* costmap, unsigned char* master_grid,
+      std::shared_ptr<std::vector<unsigned int>> shadow_points,
+      unsigned char visible_value, unsigned char shadow_value,
+      unsigned char obstacle_value) :
+        costmap_(costmap), master_grid_(master_grid),
```
Comment (dantwinkler) [inline costmap_2d/include/costmap_2d/shadow_layer.h]: Two lines

File: base_local_planner/include/base_local_planner/shadow_speed_limiter.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+private:
+  double getMapGridDistance(geometry_msgs::Point obj);
+  double distanceToVelocity(double dist);
+
+
+  costmap_2d::Costmap2D* costmap_;
+  base_local_planner::MapGrid map_grid_;
+  
+  std::vector<geometry_msgs::PoseStamped> current_pose_as_array_;
```
Comment (grieneis) [inline base_local_planner/include/base_local_planner/shadow_speed_limiter.h]: Change this.

File: base_local_planner/src/local_planner_util.cpp [hunk 10]
```diff
@@ -177,12 +178,26 @@ void LocalPlannerUtil::updateLimits() {
...
+  shadow_speed_limiter_.setShadowedObjects(costmap_->getLayeredCostmap()->getShadowedObjects());
+  shadow_speed_limiter_.setCurrentPose(robot_pose);
+  // Use the speed limiter to update the limits.
+  double sh_v_lim = 0, sh_w_lim = 0;
+  if (!shadow_speed_limiter_.calculateLimits(sh_v_lim, sh_w_lim)) {
+    ROS_WARN("Could not calculate updated speed limits.");
+    return;
+  }
+
+  ROS_INFO_THROTTLE(0.2, "Speed limits - obs: %f, shad: %f", sp_v_lim, sh_v_lim);
```
Comment (grieneis) [inline base_local_planner/src/local_planner_util.cpp]: Debug?

File: base_local_planner/src/local_planner_util.cpp [hunk 11]
```diff
@@ -177,12 +178,26 @@ void LocalPlannerUtil::updateLimits() {
   speed_limiter_.setCurrentPose(robot_pose);
 
   // Use the speed limiter to update the limits.
-  double v_lim = 0, w_lim = 0;
-  if (!speed_limiter_.calculateLimits(v_lim, w_lim)) {
+  double sp_v_lim = 0, sp_w_lim = 0;
+  if (!speed_limiter_.calculateLimits(sp_v_lim, sp_w_lim)) {
     ROS_WARN("Could not calculate updated speed limits.");
```
Comment (grieneis) [inline base_local_planner/src/local_planner_util.cpp]: Set limits to 0 and throttled error?

File: base_local_planner/src/local_planner_util.cpp [hunk 12]
```diff
@@ -177,12 +178,26 @@ void LocalPlannerUtil::updateLimits() {
...
     return;
   }
 
+  shadow_speed_limiter_.setShadowedObjects(costmap_->getLayeredCostmap()->getShadowedObjects());
+  shadow_speed_limiter_.setCurrentPose(robot_pose);
+  // Use the speed limiter to update the limits.
+  double sh_v_lim = 0, sh_w_lim = 0;
+  if (!shadow_speed_limiter_.calculateLimits(sh_v_lim, sh_w_lim)) {
+    ROS_WARN("Could not calculate updated speed limits.");
```
Comment (grieneis) [inline base_local_planner/src/local_planner_util.cpp]: Same thing.

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+  if (current_pose_as_array_.empty())
+  {
+    ROS_WARN_THROTTLE(1.0, "No pose in shadow speed limiter");
+    return true;
+  }
+
+  // calculate the brushfire grid.
+  map_grid_.resetPathDist();
+  // Adjust the pose to be at the front of the robot.
+  geometry_msgs::PoseStamped pose = current_pose_as_array_[0];
```
Comment (grieneis) [inline base_local_planner/src/shadow_speed_limiter.cpp]: Fix

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+  double pose_yaw = tf::getYaw(pose.pose.orientation);
+  pose.pose.position.x += cos(pose_yaw) * params_.forward_offset_;
+  pose.pose.position.y += sin(pose_yaw) * params_.forward_offset_;
+
```
Comment (grieneis) [inline base_local_planner/src/shadow_speed_limiter.cpp]: Maybe do with frame_id instead?

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+  // Find nearest object via brushfire distance
+
+  for (auto obj : (*objects_))
+  {
+    // Get the brushfire distance to the obstacle
+    double distance = getMapGridDistance(obj);
+    // Convert it to a velocity
+    double velocity = distanceToVelocity(distance);
+
+    max_allowed_linear_vel = (velocity < max_allowed_linear_vel) ? velocity : max_allowed_linear_vel;
```
Comment (grieneis) [inline base_local_planner/src/shadow_speed_limiter.cpp]: use min

File: base_local_planner/src/shadow_speed_limiter.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+void ShadowSpeedLimiter::setCurrentPose(tf::Stamped<tf::Pose> pose)
+{
+
+  // Covert the pose into the necessary form.
+  geometry_msgs::Pose pose_msg;
+  tf::poseTFToMsg(pose, pose_msg);
+  ROS_DEBUG("Setting current pose to %f, %f", pose_msg.position.x, pose_msg.position.y);
+  current_pose_as_array_.clear();
```
Comment (grieneis) [inline base_local_planner/src/shadow_speed_limiter.cpp]: NOt an array anymore

File: costmap_2d/include/costmap_2d/layer.h [hunk 17]
```diff
@@ -126,10 +126,18 @@ class Layer
     return false;
   }
 
+  virtual bool isShadowLayer() {
```
Comment (grieneis) [inline costmap_2d/include/costmap_2d/layer.h]: Get type -> enum class

File: costmap_2d/include/costmap_2d/shadow_layer.h [hunk 18]
```diff
@@ -0,0 +1,175 @@
+/*********************************************************************
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2008, 2013, Willow Garage, Inc.
```
Comment (grieneis) [inline costmap_2d/include/costmap_2d/shadow_layer.h]: Update

File: costmap_2d/plugins/shadow_layer.cpp [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+  ROS_DEBUG("Updating bounds for shadow layer");
+  if (rolling_window_) {
+    updateOrigin(robot_x - getSizeInMetersX() / 2, robot_y - getSizeInMetersY() / 2);
+  }
+
+  if (!enabled_) {
+    return;
+  }
+
+  // Bounds do not need to be updated b/c we are not adding cost to the costmap. yet.
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Remove comments

File: costmap_2d/plugins/shadow_layer.cpp [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+  calculateShadowedObjects();
+
+  if (publish_shadow_objects_)
+  {
+    publishShadowObjects();
+  }
+}
+
+
+// void ShadowLayer::publishShadowRegions()
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Remove

File: costmap_2d/plugins/shadow_layer.cpp [hunk 21]
```diff
@@ -0,0 +1,418 @@
...
+
+  sphere_list.type = visualization_msgs::Marker::SPHERE_LIST;
+
+
+  // POINTS markers use x and y scale for width/height respectively
+  sphere_list.scale.x = min_shadow_size_;
+  sphere_list.scale.y = min_shadow_size_;
+  sphere_list.scale.z = min_shadow_size_;
+
+  // Points are green
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Fix comment

File: costmap_2d/plugins/shadow_layer.cpp [hunk 22]
```diff
@@ -0,0 +1,418 @@
...
+    double wy = cloud.points[i].y;
+
+    double a = wx - ox;
+    double b = wy - oy;
+    double ray_length = sqrt(a * a + b * b);
+
+
+    // calculate raytrace starting point
+    // the raytrace range should be (rx, ry) -> (wx, wy)
+    double rx = ox;
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Clean up.

File: costmap_2d/plugins/shadow_layer.cpp [hunk 23]
```diff
@@ -0,0 +1,418 @@
...
+{
+
+  if (!shadowed_objects_)
+  {
+    shadowed_objects_ = std::make_shared<std::vector<geometry_msgs::Point>>();
+  }
+  shadowed_objects_->clear();
+  ROS_DEBUG("Checking %d shadowed points", shadowed_points_->size());
+  // Iterate over all of the shadowed_points_
+  for (auto it : (*shadowed_points_))
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: const auto & it

File: costmap_2d/plugins/shadow_layer.cpp [hunk 24]
```diff
@@ -0,0 +1,418 @@
...
+  for (unsigned int y_val = start_y; y_val <= end_y; ++y_val)
+  {
+    for (unsigned int x_val = start_x; x_val <= end_x; ++x_val)
+    {
+      switch (costmap_[getIndex(x_val, y_val)])
+      {
+        case VISIBLE:
+        case OBSTACLE:
+          return false;
+        default:
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Remove?

File: costmap_2d/plugins/shadow_layer.cpp [hunk 25]
```diff
@@ -0,0 +1,418 @@
...
+geometry_msgs::Point ShadowLayer::createPointFromIndex(unsigned int idx)
+{
+  geometry_msgs::Point pt;
+  unsigned int mx, my;
+  indexToCells(idx, mx, my);
+  mapToWorld(mx, my, pt.x, pt.y);
+  return pt;
+}
+
+void ShadowLayer::updateRaytraceBounds(double ox, double oy, double wx, double wy, double range,
```
Comment (grieneis) [inline costmap_2d/plugins/shadow_layer.cpp]: Remove?

File: costmap_2d/include/costmap_2d/shadow_layer.h (hunk 1 above)
Comment (grieneis) [reply costmap_2d/include/costmap_2d/shadow_layer.h]: Done

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 2 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 3 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Good to know.  Done.

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 6 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Done

File: costmap_2d/cfg/ShadowPlugin.cfg (hunk 7 above)
Comment (grieneis) [reply costmap_2d/cfg/ShadowPlugin.cfg]: Done

File: costmap_2d/include/costmap_2d/shadow_layer.h (hunk 8 above)
Comment (grieneis) [reply costmap_2d/include/costmap_2d/shadow_layer.h]: Done

File: base_local_planner/include/base_local_planner/shadow_speed_limiter.h (hunk 9 above)
Comment (grieneis) [reply base_local_planner/include/base_local_planner/shadow_speed_limiter.h]: Done

File: base_local_planner/src/local_planner_util.cpp (hunk 11 above)
Comment (grieneis) [reply base_local_planner/src/local_planner_util.cpp]: Done

File: base_local_planner/src/local_planner_util.cpp (hunk 12 above)
Comment (grieneis) [reply base_local_planner/src/local_planner_util.cpp]: Done

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 13 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Done

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 14 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Skipping for now

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 15 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Done

File: base_local_planner/src/shadow_speed_limiter.cpp (hunk 16 above)
Comment (grieneis) [reply base_local_planner/src/shadow_speed_limiter.cpp]: Done

File: costmap_2d/include/costmap_2d/layer.h (hunk 17 above)
Comment (grieneis) [reply costmap_2d/include/costmap_2d/layer.h]: Done

File: costmap_2d/include/costmap_2d/shadow_layer.h (hunk 18 above)
Comment (grieneis) [reply costmap_2d/include/costmap_2d/shadow_layer.h]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 19 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 20 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 21 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 22 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Actually looks good.

File: costmap_2d/plugins/shadow_layer.cpp (hunk 23 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp (hunk 25 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: base_local_planner/src/local_planner_util.cpp [hunk 26]
```diff
@@ -173,30 +171,37 @@ void LocalPlannerUtil::updateLimits() {
   tf::Stamped<tf::Pose> robot_pose;
   if (!costmap_->getRobotPose(robot_pose)) {
     ROS_WARN("Could not get robot pose to calculate speed limits");
-    return;
+    return false;
```
Comment (dantwinkler) [inline base_local_planner/src/local_planner_util.cpp:174]: Good catch! :)

File: costmap_2d/include/costmap_2d/layer.h [hunk 27]
```diff
@@ -118,17 +126,9 @@ class Layer
...
-
-  virtual bool isObstructionLayer() {
-    return false;
-  }
-
-  virtual bool isShadowLayer() {
-    return false;
-  }
+  virtual LayerType getLayerType() {
+    return LayerType::UNKNOWN;
```
Comment (dantwinkler) [inline costmap_2d/include/costmap_2d/layer.h]: Is it possible to create the base layer without a type?  If not we should we make this pure virtual?  If so should unknown but default.

**Review by dantwinkler — APPROVED** [review]

File: costmap_2d/plugins/shadow_layer.cpp [hunk 28]
```diff
@@ -0,0 +1,358 @@
...
+    double ry = oy;
+
+    // now we need to make sure that the point we're raytracing
+    // to isn't off the costmap and scale if necessary
+    checkRaytracePoint(origin_x, origin_y, map_end_x, map_end_y, ox, oy, wx, wy);
+    // now that the vector is scaled correctly... we'll get the map coordinates of its endpoint
+    unsigned int x1, y1;
+
+    // check for legality just in case
+    if (!worldToMap(wx, wy, x1, y1))
```
Comment (mrousseau65) [inline costmap_2d/plugins/shadow_layer.cpp]: Missing braces

Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Done

File: costmap_2d/plugins/shadow_layer.cpp [hunk 29]
```diff
@@ -0,0 +1,358 @@
...
+
+  // for each point in the cloud, we want to trace a line from the origin and clear obstacles along it
+  for (unsigned int i = 0; i < cloud.points.size(); ++i)
+  {
+    double wx = cloud.points[i].x;
+    double wy = cloud.points[i].y;
+
+    double a = wx - ox;
+    double b = wy - oy;
+    double ray_length = sqrt(a * a + b * b);
```
Comment (mrousseau65) [inline costmap_2d/plugins/shadow_layer.cpp]: Is this used anywhere (if it is, hypot may be cheaper)?

File: costmap_2d/plugins/shadow_layer.cpp [hunk 30]
```diff
@@ -0,0 +1,358 @@
...
+  shadowed_points_->clear();
+
+  // Check if the observation is valid.
+  if (!shadow_observation_)
+  {
+    ROS_WARN_THROTTLE(1.0, "Cannot raytrace in shadow layer because there is no shadow observation.");
+    return;
+  }
+
+  //
```
Comment (mrousseau65) [inline costmap_2d/plugins/shadow_layer.cpp]: ?

Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Hmmm...

File: costmap_2d/plugins/shadow_layer.cpp (hunk 29 above)
Comment (grieneis) [reply costmap_2d/plugins/shadow_layer.cpp]: Good catch - completely unused and wasting cycles.

**Review by dantwinkler — APPROVED** [review]

**Pull request merged**